File: backend/services/providers/yahoo.py
"""
Yahoo Finance provider — uses the v8 chart API directly (no yfinance).
No API key required. Rate limit: ~5–10 req/s per IP in practice.
Safe polling floor: ~15s per ticker.
"""
import requests
from datetime import datetime
from .base import DataProvider


class YahooProvider(DataProvider):

    TTL = {
        "quote":      15,
        "history":    60,
        "news":      120,
        "financials": 3600,
        "options":    300,
        "screen":      30,
    }

    def __init__(self):
        self._session = requests.Session()
        self._session.headers.update({
            "User-Agent": (
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/120.0.0.0 Safari/537.36"
            ),
            "Accept": "application/json",
            "Accept-Language": "en-US,en;q=0.9",
        })

    # ── Internal helpers ──────────────────────────────────────────────────────

    def _chart(self, symbol: str, interval: str = "1d", range_: str = "1mo"):
        url = f"https://query1.finance.yahoo.com/v8/finance/chart/{symbol}"
        params = {"interval": interval, "range": range_, "includePrePost": "false"}
        try:
            r = self._session.get(url, params=params, timeout=12)
            if not r.ok:
                return None
            results = r.json().get("chart", {}).get("result")
            return results[0] if results else None
        except Exception:
            return None

    def _quote_summary(self, symbol: str, modules: str):
        url = f"https://query1.finance.yahoo.com/v11/finance/quoteSummary/{symbol}"
        try:
            r = self._session.get(url, params={"modules": modules, "formatted": "false"}, timeout=12)
            if not r.ok:
                return None
            result = r.json().get("quoteSummary", {}).get("result")
            return result[0] if result else None
        except Exception:
            return None

    # ── DataProvider interface ────────────────────────────────────────────────

    def get_quote(self, ticker: str) -> dict:
        result = self._chart(ticker, interval="1d", range_="5d")
        if not result:
            return {"error": f"No data for {ticker}. Check the ticker symbol."}

        meta = result.get("meta", {})
        price = meta.get("regularMarketPrice")
        prev_close = meta.get("chartPreviousClose") or meta.get("previousClose")

        if price is None:
            return {"error": f"No price data for {ticker}."}

        change = round(price - prev_close, 4) if prev_close else 0
        change_pct = round((change / prev_close) * 100, 4) if prev_close else 0

        info = {}
        qs = self._quote_summary(
            ticker,
            "assetProfile,defaultKeyStatistics,summaryDetail,financialData,price",
        )
        if qs:
            info = {
                **(qs.get("assetProfile") or {}),
                **(qs.get("defaultKeyStatistics") or {}),
                **(qs.get("summaryDetail") or {}),
                **(qs.get("financialData") or {}),
                **(qs.get("price") or {}),
            }

        return {
            "symbol":        ticker.upper(),
            "name":          info.get("longName") or info.get("shortName") or meta.get("longName") or meta.get("shortName") or ticker.upper(),
            "price":         round(float(price), 2),
            "change":        round(float(change), 2),
            "change_pct":    round(float(change_pct), 2),
            "volume":        int(meta.get("regularMarketVolume") or info.get("volume") or 0),
            "avg_volume":    int(info.get("averageVolume") or info.get("averageDailyVolume10Day") or 0),
            "market_cap":    info.get("marketCap") or meta.get("marketCap"),
            "pe_ratio":      info.get("trailingPE") or info.get("forwardPE"),
            "eps":           info.get("trailingEps"),
            "week_52_high":  meta.get("fiftyTwoWeekHigh") or info.get("fiftyTwoWeekHigh"),
            "week_52_low":   meta.get("fiftyTwoWeekLow") or info.get("fiftyTwoWeekLow"),
            "dividend_yield":info.get("dividendYield") or info.get("trailingAnnualDividendYield"),
            "beta":          info.get("beta"),
            "sector":        info.get("sector"),
            "industry":      info.get("industry"),
            "description":   info.get("longBusinessSummary", ""),
            "market_status": meta.get("marketState", "").lower(),
            "exchange":      meta.get("exchangeName") or meta.get("fullExchangeName") or "",
            "currency":      meta.get("currency", "USD"),
        }

    def get_history(self, ticker: str, period: str, interval: str) -> list:
        period_map = {
            "1d": "1d", "5d": "5d", "1mo": "1mo", "3mo": "3mo",
            "6mo": "6mo", "1y": "1y", "2y": "2y", "5y": "5y", "max": "max",
        }
        result = self._chart(ticker, interval=interval, range_=period_map.get(period, "1mo"))
        if not result:
            return []

        timestamps = result.get("timestamp", [])
        q = result.get("indicators", {}).get("quote", [{}])[0]
        opens, highs, lows, closes, volumes = (
            q.get("open", []), q.get("high", []), q.get("low", []),
            q.get("close", []), q.get("volume", []),
        )

        out = []
        for i, ts in enumerate(timestamps):
            o, h, l, c = (
                opens[i]  if i < len(opens)  else None,
                highs[i]  if i < len(highs)  else None,
                lows[i]   if i < len(lows)   else None,
                closes[i] if i < len(closes) else None,
            )
            if None in (o, h, l, c):
                continue
            v  = volumes[i] if i < len(volumes) else 0
            dt = datetime.utcfromtimestamp(ts)
            out.append({
                "time":   dt.strftime("%Y-%m-%d") if interval in ("1d", "1wk", "1mo") else int(ts),
                "open":   round(float(o), 4),
                "high":   round(float(h), 4),
                "low":    round(float(l), 4),
                "close":  round(float(c), 4),
                "volume": int(v or 0),
            })
        return out

    def get_news(self, ticker: str) -> list:
        url = "https://query1.finance.yahoo.com/v1/finance/search"
        try:
            r = self._session.get(
                url,
                params={"q": ticker, "newsCount": 20, "enableFuzzyQuery": False, "quotesCount": 0},
                timeout=12,
            )
            if not r.ok:
                return []
            out = []
            for item in r.json().get("news", []):
                ts = item.get("providerPublishTime")
                pub = ""
                if ts:
                    try:
                        pub = datetime.utcfromtimestamp(ts).strftime("%Y-%m-%d %H:%M")
                    except Exception:
                        pub = str(ts)
                out.append({
                    "title":        item.get("title", ""),
                    "publisher":    item.get("publisher", ""),
                    "link":         item.get("link", ""),
                    "published_at": pub,
                    "sentiment":    None,
                })
            return out
        except Exception:
            return []

    def get_financials(self, ticker: str, period: str = "annual") -> dict:
        if period == "quarterly":
            modules = "incomeStatementHistoryQuarterly,balanceSheetHistoryQuarterly,cashflowStatementHistoryQuarterly"
        else:
            modules = "incomeStatementHistory,balanceSheetHistory,cashflowStatementHistory"

        qs = self._quote_summary(ticker, modules)
        if not qs:
            return {"error": "Could not fetch financials"}

        def _extract(section_key):
            section = qs.get(section_key, {})
            # Each section has one list key containing the statements
            items = next((v for v in section.values() if isinstance(v, list)), [])
            out = {}
            for stmt in items:
                date = stmt.get("endDate", {})
                date_str = date.get("fmt", str(date)) if isinstance(date, dict) else str(date)
                row = {}
                for k, v in stmt.items():
                    if k in ("endDate", "maxAge"):
                        continue
                    row[k] = v.get("raw") if isinstance(v, dict) else v
                out[date_str] = row
            return out

        if period == "quarterly":
            return {
                "income":   _extract("incomeStatementHistoryQuarterly"),
                "balance":  _extract("balanceSheetHistoryQuarterly"),
                "cashflow": _extract("cashflowStatementHistoryQuarterly"),
            }
        return {
            "income":   _extract("incomeStatementHistory"),
            "balance":  _extract("balanceSheetHistory"),
            "cashflow": _extract("cashflowStatementHistory"),
        }

    def get_options(self, ticker: str) -> dict:
        url = f"https://query1.finance.yahoo.com/v7/finance/options/{ticker}"
        try:
            r = self._session.get(url, timeout=12)
            if not r.ok:
                return {"error": "Options data unavailable"}
            opt_result = r.json().get("optionChain", {}).get("result", [])
            if not opt_result:
                return {"error": "No options data"}

            out = {}
            for exp_ts in opt_result[0].get("expirationDates", [])[:3]:
                exp_str = datetime.utcfromtimestamp(exp_ts).strftime("%Y-%m-%d")
                r2 = self._session.get(url, params={"date": exp_ts}, timeout=12)
                if not r2.ok:
                    continue
                r2_res = r2.json().get("optionChain", {}).get("result", [])
                if not r2_res:
                    continue
                options = r2_res[0].get("options", [{}])[0]

                def _clean(contracts):
                    def _raw(v):
                        return v.get("raw") if isinstance(v, dict) else v
                    return [{
                        "strike": _raw(c.get("strike")),
                        "last":   _raw(c.get("lastPrice")),
                        "bid":    _raw(c.get("bid")),
                        "ask":    _raw(c.get("ask")),
                        "volume": _raw(c.get("volume")),
                        "oi":     _raw(c.get("openInterest")),
                        "iv":     round(float(_raw(c.get("impliedVolatility")) or 0) * 100, 1),
                    } for c in contracts[:20]]

                out[exp_str] = {
                    "calls": _clean(options.get("calls", [])),
                    "puts":  _clean(options.get("puts", [])),
                }
            return out
        except Exception as e:
            return {"error": str(e)}

    def get_screen(self, mode: str = "active") -> list:
        scr_id = {"active": "most_actives", "gainers": "day_gainers", "losers": "day_losers"}.get(mode, "most_actives")
        url = (
            "https://query1.finance.yahoo.com/v1/finance/screener/predefined/saved"
            f"?formatted=false&scrIds={scr_id}&count=25"
        )
        try:
            r = self._session.get(url, timeout=12)
            if not r.ok:
                return []
            quotes = r.json().get("finance", {}).get("result", [{}])[0].get("quotes", [])
            return [{
                "symbol":     q.get("symbol", ""),
                "name":       q.get("longName") or q.get("shortName") or "",
                "price":      q.get("regularMarketPrice"),
                "change":     round(q.get("regularMarketChange") or 0, 2),
                "change_pct": round(q.get("regularMarketChangePercent") or 0, 2),
                "volume":     q.get("regularMarketVolume"),
            } for q in quotes]
        except Exception:
            return []

File: backend-rust/src/yahoo.rs
use reqwest::Client;
use serde_json::{json, Value};
use std::time::Duration;

pub fn make_client() -> Client {
    use reqwest::header::{self, HeaderMap, HeaderValue};
    let mut headers = HeaderMap::new();
    headers.insert(
        header::USER_AGENT,
        HeaderValue::from_static(
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) \
             AppleWebKit/537.36 (KHTML, like Gecko) \
             Chrome/120.0.0.0 Safari/537.36",
        ),
    );
    headers.insert(
        header::ACCEPT_LANGUAGE,
        HeaderValue::from_static("en-US,en;q=0.9"),
    );
    headers.insert(
        header::ACCEPT,
        HeaderValue::from_static("application/json"),
    );
    Client::builder()
        .default_headers(headers)
        .timeout(Duration::from_secs(12))
        .build()
        .unwrap()
}

// ── Helpers ───────────────────────────────────────────────────────────────────

fn f64_val(v: &Value) -> Option<f64> {
    match v {
        Value::Number(n) => n.as_f64(),
        Value::String(s) => s.parse().ok(),
        _ => None,
    }
}

fn get_f64(v: &Value, key: &str) -> Option<f64> {
    f64_val(v.get(key)?)
}

fn get_i64(v: &Value, key: &str) -> Option<i64> {
    v.get(key)?.as_i64()
}

fn get_str<'a>(v: &'a Value, key: &str) -> &'a str {
    v.get(key).and_then(|x| x.as_str()).unwrap_or("")
}

fn coalesce_f64(v: &Value, keys: &[&str]) -> Option<f64> {
    keys.iter().find_map(|k| get_f64(v, k))
}

fn coalesce_str<'a>(v: &'a Value, keys: &[&str]) -> &'a str {
    keys.iter()
        .find_map(|k| v.get(k).and_then(|x| x.as_str()).filter(|s| !s.is_empty()))
        .unwrap_or("")
}

// ── Yahoo Finance API calls ───────────────────────────────────────────────────

async fn chart(client: &Client, symbol: &str, interval: &str, range: &str) -> Option<Value> {
    let url = format!("https://query1.finance.yahoo.com/v8/finance/chart/{symbol}");
    let res = client
        .get(&url)
        .query(&[("interval", interval), ("range", range), ("includePrePost", "false")])
        .send()
        .await
        .ok()?;
    if !res.status().is_success() { return None; }
    let body: Value = res.json().await.ok()?;
    let results = body.get("chart")?.get("result")?.as_array()?;
    results.first().cloned()
}

async fn quote_summary(client: &Client, symbol: &str, modules: &str) -> Option<Value> {
    let url = format!("https://query1.finance.yahoo.com/v11/finance/quoteSummary/{symbol}");
    let res = client
        .get(&url)
        .query(&[("modules", modules), ("formatted", "false")])
        .send()
        .await
        .ok()?;
    if !res.status().is_success() { return None; }
    let body: Value = res.json().await.ok()?;
    let results = body.get("quoteSummary")?.get("result")?.as_array()?;
    results.first().cloned()
}

// ── Public provider methods ───────────────────────────────────────────────────

pub async fn get_quote(client: &Client, ticker: &str) -> Option<Value> {
    let result = chart(client, ticker, "1d", "5d").await?;
    let meta = result.get("meta")?;

    let price = get_f64(meta, "regularMarketPrice")?;
    let prev_close = coalesce_f64(meta, &["chartPreviousClose", "previousClose"]).unwrap_or(price);
    let change = ((price - prev_close) * 10000.0).round() / 10000.0;
    let change_pct = if prev_close != 0.0 {
        ((change / prev_close) * 1_000_000.0).round() / 10000.0
    } else { 0.0 };

    // Merge all quoteSummary modules into one flat map
    let qs = quote_summary(
        client, ticker,
        "assetProfile,defaultKeyStatistics,summaryDetail,financialData,price",
    ).await;

    let info = qs.as_ref().map(|q| {
        let mut merged = serde_json::Map::new();
        for module in &["assetProfile","defaultKeyStatistics","summaryDetail","financialData","price"] {
            if let Some(Value::Object(map)) = q.get(module) {
                merged.extend(map.clone());
            }
        }
        Value::Object(merged)
    }).unwrap_or(json!({}));

    let name = coalesce_str(&info, &["longName","shortName"]);
    let name = if name.is_empty() {
        coalesce_str(meta, &["longName","shortName"])
    } else { name };
    let name = if name.is_empty() { ticker } else { name };

    Some(json!({
        "symbol":         ticker.to_uppercase(),
        "name":           name,
        "price":          (price * 100.0).round() / 100.0,
        "change":         (change * 100.0).round() / 100.0,
        "change_pct":     (change_pct * 100.0).round() / 100.0,
        "volume":         get_i64(meta, "regularMarketVolume").or_else(|| get_i64(&info, "volume")).unwrap_or(0),
        "avg_volume":     coalesce_f64(&info, &["averageVolume","averageDailyVolume10Day"]).map(|v| v as i64).unwrap_or(0),
        "market_cap":     coalesce_f64(&info, &["marketCap"]).or_else(|| get_f64(meta, "marketCap")),
        "pe_ratio":       coalesce_f64(&info, &["trailingPE","forwardPE"]),
        "eps":            get_f64(&info, "trailingEps"),
        "week_52_high":   coalesce_f64(meta, &["fiftyTwoWeekHigh"]).or_else(|| get_f64(&info, "fiftyTwoWeekHigh")),
        "week_52_low":    coalesce_f64(meta, &["fiftyTwoWeekLow"]).or_else(|| get_f64(&info, "fiftyTwoWeekLow")),
        "dividend_yield": coalesce_f64(&info, &["dividendYield","trailingAnnualDividendYield"]),
        "beta":           get_f64(&info, "beta"),
        "sector":         get_str(&info, "sector"),
        "industry":       get_str(&info, "industry"),
        "description":    get_str(&info, "longBusinessSummary"),
        "market_status":  get_str(meta, "marketState").to_lowercase(),
        "exchange":       coalesce_str(meta, &["exchangeName","fullExchangeName"]),
        "currency":       coalesce_str(meta, &["currency"]),
    }))
}

pub async fn get_history(client: &Client, ticker: &str, period: &str, interval: &str) -> Vec<Value> {
    let result = match chart(client, ticker, interval, period).await {
        Some(r) => r,
        None => return vec![],
    };

    let timestamps = match result.get("timestamp").and_then(|t| t.as_array()) {
        Some(t) => t.clone(),
        None => return vec![],
    };

    let q = result
        .get("indicators")
        .and_then(|i| i.get("quote"))
        .and_then(|q| q.as_array())
        .and_then(|a| a.first())
        .cloned()
        .unwrap_or(json!({}));

    let empty = vec![];
    let opens   = q.get("open").and_then(|v| v.as_array()).unwrap_or(&empty);
    let highs   = q.get("high").and_then(|v| v.as_array()).unwrap_or(&empty);
    let lows    = q.get("low").and_then(|v| v.as_array()).unwrap_or(&empty);
    let closes  = q.get("close").and_then(|v| v.as_array()).unwrap_or(&empty);
    let volumes = q.get("volume").and_then(|v| v.as_array()).unwrap_or(&empty);

    let daily = matches!(interval, "1d" | "1wk" | "1mo");

    timestamps.iter().enumerate().filter_map(|(i, ts)| {
        let ts_i64 = ts.as_i64()?;
        let o = f64_val(opens.get(i)?)?;
        let h = f64_val(highs.get(i)?)?;
        let l = f64_val(lows.get(i)?)?;
        let c = f64_val(closes.get(i)?)?;
        let v = volumes.get(i).and_then(|x| x.as_i64()).unwrap_or(0);

        let time_val = if daily {
            let dt = chrono::DateTime::from_timestamp(ts_i64, 0)?;
            json!(dt.format("%Y-%m-%d").to_string())
        } else {
            json!(ts_i64)
        };

        Some(json!({
            "time":   time_val,
            "open":   (o * 10000.0).round() / 10000.0,
            "high":   (h * 10000.0).round() / 10000.0,
            "low":    (l * 10000.0).round() / 10000.0,
            "close":  (c * 10000.0).round() / 10000.0,
            "volume": v,
        }))
    }).collect()
}

pub async fn get_news(client: &Client, ticker: &str) -> Vec<Value> {
    let url = "https://query1.finance.yahoo.com/v1/finance/search";
    let res = match client
        .get(url)
        .query(&[("q", ticker), ("newsCount", "20"), ("enableFuzzyQuery", "false"), ("quotesCount", "0")])
        .send()
        .await
    {
        Ok(r) if r.status().is_success() => r,
        _ => return vec![],
    };

    let body: Value = match res.json().await {
        Ok(v) => v,
        Err(_) => return vec![],
    };

    body.get("news")
        .and_then(|n| n.as_array())
        .map(|items| {
            items.iter().map(|item| {
                let ts = item.get("providerPublishTime").and_then(|t| t.as_i64());
                let published_at = ts
                    .and_then(|t| chrono::DateTime::from_timestamp(t, 0))
                    .map(|dt| dt.format("%Y-%m-%d %H:%M").to_string())
                    .unwrap_or_default();
                json!({
                    "title":        get_str(item, "title"),
                    "publisher":    get_str(item, "publisher"),
                    "link":         get_str(item, "link"),
                    "published_at": published_at,
                    "sentiment":    null,
                })
            }).collect()
        })
        .unwrap_or_default()
}

pub async fn get_financials(client: &Client, ticker: &str, period: &str) -> Option<Value> {
    let modules = if period == "quarterly" {
        "incomeStatementHistoryQuarterly,balanceSheetHistoryQuarterly,cashflowStatementHistoryQuarterly"
    } else {
        "incomeStatementHistory,balanceSheetHistory,cashflowStatementHistory"
    };

    let qs = quote_summary(client, ticker, modules).await?;

    let extract = |section_key: &str| -> Value {
        let section = match qs.get(section_key).and_then(|s| s.as_object()) {
            Some(s) => s,
            None => return json!({}),
        };
        // Each section has one list key containing the statements
        let stmts = section.values()
            .find_map(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let mut out = serde_json::Map::new();
        for stmt in stmts {
            let date = stmt.get("endDate")
                .and_then(|d| d.get("fmt"))
                .and_then(|f| f.as_str())
                .unwrap_or("")
                .to_string();
            if date.is_empty() { continue; }

            let mut row = serde_json::Map::new();
            if let Some(obj) = stmt.as_object() {
                for (k, v) in obj {
                    if k == "endDate" || k == "maxAge" { continue; }
                    let raw = v.get("raw").cloned().unwrap_or(v.clone());
                    row.insert(k.clone(), raw);
                }
            }
            out.insert(date, Value::Object(row));
        }
        Value::Object(out)
    };

    let (income_key, balance_key, cashflow_key) = if period == "quarterly" {
        (
            "incomeStatementHistoryQuarterly",
            "balanceSheetHistoryQuarterly",
            "cashflowStatementHistoryQuarterly",
        )
    } else {
        (
            "incomeStatementHistory",
            "balanceSheetHistory",
            "cashflowStatementHistory",
        )
    };

    Some(json!({
        "income":   extract(income_key),
        "balance":  extract(balance_key),
        "cashflow": extract(cashflow_key),
    }))
}

pub async fn get_screen(client: &Client, mode: &str) -> Vec<Value> {
    let scr_id = match mode {
        "gainers" => "day_gainers",
        "losers"  => "day_losers",
        _         => "most_actives",
    };

    let url = format!(
        "https://query1.finance.yahoo.com/v1/finance/screener/predefined/saved\
         ?formatted=false&scrIds={scr_id}&count=25"
    );

    let res = match client.get(&url).send().await {
        Ok(r) if r.status().is_success() => r,
        _ => return vec![],
    };

    let body: Value = match res.json().await {
        Ok(v) => v,
        Err(_) => return vec![],
    };

    body.get("finance")
        .and_then(|f| f.get("result"))
        .and_then(|r| r.as_array())
        .and_then(|a| a.first())
        .and_then(|r| r.get("quotes"))
        .and_then(|q| q.as_array())
        .map(|quotes| {
            quotes.iter().map(|q| json!({
                "symbol":     get_str(q, "symbol"),
                "name":       coalesce_str(q, &["longName","shortName"]),
                "price":      get_f64(q, "regularMarketPrice"),
                "change":     get_f64(q, "regularMarketChange").map(|v| (v * 100.0).round() / 100.0),
                "change_pct": get_f64(q, "regularMarketChangePercent").map(|v| (v * 100.0).round() / 100.0),
                "volume":     get_i64(q, "regularMarketVolume"),
            })).collect()
        })
        .unwrap_or_default()
}
